### PR TITLE
[hail] clean up generated IR for struct.select() with only named fields

### PR DESCRIPTION
**Pull request by catoverdrive**
We're currently generating IR that looks like:
```
(InsertFields
  (SelectFields row ())
  (f1 …)
  …
)
```
for something like row.select(**new_field_exprs).

I did this in the process of debugging #7638, but it doesn't actually fix the issue completely; I'm getting past the stack overflow issue on a table with 2000 fields, but then running into a class size issue when trying to generate the code.

### CODE REVIEW

**Comment by tpoterba** [comment]
will take this since Chris is on vacation

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 1]
```diff
@@ -1623,6 +1623,9 @@ def select(self, *fields, **named_exprs):
             Struct containing specified existing fields and computed fields.
         """
 
+        if len(fields) == 0:
+            return hl.or_missing(hl.is_defined(self), hl.struct(**named_exprs))
```
Comment (tpoterba) [inline hail/python/hail/expr/expressions/typed_expressions.py]: Is this actually an improvement?

I think one easy optimization we should be doing (but aren't currently) in the current IR is to rewrite this:
```
ht.row.select(**ht.row_key, **other_fields)
```
to:
```
ht.row.select(*ht.row_key, **other_fields)
```

Which can be implemented more efficiently (with the right ptype).

Comment (tpoterba) [reply hail/python/hail/expr/expressions/typed_expressions.py]: Yours will generate a Let and an If, to replace a SelectFields.

Comment (catoverdrive) [reply hail/python/hail/expr/expressions/typed_expressions.py]: I don't think it's strictly an improvement if we look at the code being generated from the IR, but it does prevent the optimization passes in a Table.show from overflowing the stack. (I think we have more expensive introspection on how InsertFields and SelectFields interacts with bindings, so this at least will help us not do that.)

Comment (tpoterba) [reply hail/python/hail/expr/expressions/typed_expressions.py]: I really think this is a deoptimization.

I think the core problem is that TableShow is implemented in Python and generates HUMONGOUS IRs.

Comment (tpoterba) [reply hail/python/hail/expr/expressions/typed_expressions.py]: can you get the show to run if you, say, quadruple the stack size?

Comment (tpoterba) [reply hail/python/hail/expr/expressions/typed_expressions.py]: (setting java opts)

**Pull request closed without merging**